services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,392 @@
+
+        return portStateJson;
+    }
+
+    private RouterState getRouterState(String routerId) {
+        return null;
+    }
+
+    private void verifyRouter(String deviceId, String tenantId) throws Exception {
+        RouterState routerState = getRouterState(deviceId);
+
+        if (routerState == null) {
+            throw new RouterNotFoundException();
+        }
+
+        if (!tenantId.equals(routerState.getTenantId())) {
+            throw new RouterNotOwnedByTenant();
+        }
+    }
+
+    private Map<String, Set<String>> fixedIpsToMap(List<PortState.FixedIp> fixedIps) {
+        Map<String, Set<String>> subnetIpsMap = new HashMap<>();
+
+        for (PortState.FixedIp fixedIp: fixedIps) {
+            if (subnetIpsMap.containsKey(fixedIp.getSubnetId())) {
+                subnetIpsMap.get(fixedIp.getSubnetId()).add(fixedIp.getIpAddress());
+            } else {
+                Set<String> ips = new HashSet<>();
+                ips.add(fixedIp.getIpAddress());
+                subnetIpsMap.put(fixedIp.getSubnetId(), ips);
+            }
+        }
+
+        return subnetIpsMap;
+    }
+
+    private List<PortState.FixedIp> fixedIpsCompare(List<PortState.FixedIp> fixedIps1, List<PortState.FixedIp> fixedIps2) {
+        List<PortState.FixedIp> addFixedIps = new ArrayList<>();
+        Map<String, Set<String>> subnetIpsMap = fixedIpsToMap(fixedIps2);
+
+        for (PortState.FixedIp fixedIp: fixedIps1) {
+            String subnetId = fixedIp.getSubnetId();
+            String ipAddress = fixedIp.getIpAddress();
+            if (subnetIpsMap.containsKey(subnetId)) {
+                if (!subnetIpsMap.get(subnetId).contains(ipAddress)) {
+                    addFixedIps.add(fixedIp);
+                }
+            } else {
+                addFixedIps.add(fixedIp);
+            }
+        }
+
+        return addFixedIps;
+    }
+
+    private void updateSecurityGroup(PortState portState, PortState oldPortState) throws Exception {
+        String deviceOwner = portState.getDeviceOwner();
+
+        //Network device interface does not need security groups
+        if (deviceOwner != null && deviceOwner.indexOf("network") > 0) {
+            throw new UpdateSecurityGroupException();
+        }
+
+        //Verify request security groups valid
+        verifySecurityGroup(portState);
+
+        //Delete old security groups binding
+        unbindSecurityGroups(oldPortState);
+
+        //Create security groups binding for port
+        bindSecurityGroups(portState);
+
+        oldPortState.setSecurityGroups(portState.getSecurityGroups());
+    }
+
+    private void UpdateExtraDhcpOpts(PortState portState, PortState portStateOld) {
+
+    }
+
+    private void updatePortToHost(PortState portState) {
+
+    }
+
+    public PortStateJson updatePortState(String projectId, String portId, PortStateJson portStateJson) throws Exception {
```

Comment:
Looks like we'll need a bit more time for updatePortState. It is fine and let us do it in next PR.